### PR TITLE
Rewrite top level `this` to `undefined` in Harmony modules

### LINKED ISSUE #5074: This in module not undefined
**Do you want to request a *feature* or report a *bug*?**
Report a bug.

**What is the current behavior?**
The value of `this` at the top level in a module is not `undefined`; instead it's an object.
**If the current behavior is a bug, please provide the steps to reproduce.**
```js
export default 4;

console.log('module this', this);
```

**What is the expected behavior?**
`this` at the top level in the module must be `undefined`: https://tc39.github.io/ecma262/#sec-module-environment-records-getthisbinding
**If this is a feature request, what is motivation or use case for changing the behavior?**

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**
webpack 2.6.1, node 6.11.0, ubuntu

### PR DESCRIPTION
**Pull request by ooflorent**
**What kind of change does this PR introduce?**

bugfix

**Did you add tests for your changes?**

yes

**If relevant, link to documentation update:**

n/a

**Summary**

Rewrite `this` to `undefined` in Harmony modules to be [spec-compliant](https://tc39.github.io/ecma262/#sec-module-environment-records-getthisbinding).

> **8.1.1.5.4 `GetThisBinding()`**
> 1. Return `undefined`.

```js
// in
console.log(this, this.exports);
export default this;

// out
console.log(undefined, undefined.exports);
export default undefined;
```

Fixes #5074 

**Does this PR introduce a breaking change?**

Yes if a module was relying on this broken behavior.

### CODE REVIEW

File: lib/dependencies/HarmonyTopLevelThisParserPlugin.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Florent Cailhol @ooflorent
+*/
+"use strict";
+
+const walk = require("acorn/dist/walk");
+const ConstDependency = require("./ConstDependency");
+
+const rewriteThisVisitor = {
```
Comment (sokra) [inline lib/dependencies/HarmonyTopLevelThisParserPlugin.js]: Instead of a custom walker in the plugin, add the ability to parse top-level-this to the Parser with a new hook i. e. `topLevelThis`.

File: lib/dependencies/HarmonyTopLevelThisParserPlugin.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Florent Cailhol @ooflorent
+*/
+"use strict";
+
+const ConstDependency = require("./ConstDependency");
+
+class HarmonyTopLevelThisParserPlugin {
+	apply(parser) {
+		let isHarmony = false;
```
Comment (Kovensky) [inline lib/dependencies/HarmonyTopLevelThisParserPlugin.js]: Is there a separate instance of `(new HarmonyTopLevelThisParserPlugin()).apply` for each module, or is this taking advantage of a race condition?

Comment (ooflorent) [reply lib/dependencies/HarmonyTopLevelThisParserPlugin.js]: > is this taking advantage of a race condition?

Yes. It reads metadata set by `HarmonyDetectionParserPlugin`.

File: lib/dependencies/HarmonyTopLevelThisParserPlugin.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+		parser.hooks.program.tap("HarmonyTopLevelThisParserPlugin", () => {
```
Comment (sokra) [inline lib/dependencies/HarmonyTopLevelThisParserPlugin.js]: This hook is not needed. You can move the isHarmony logic into the `this expression` hook

File: lib/Parser.js [hunk 4]
```diff
@@ -624,10 +624,15 @@ class Parser extends Tapable {
 	}
 
 	walkMethodDefinition(methodDefinition) {
+		const wasInFunction = this.scope.inFunction;
+		this.scope.inFunction = true;
```
Comment (sokra) [inline lib/Parser.js]: Maybe renamed this to `topLevelScope` instead of `inFunction`

File: lib/dependencies/HarmonyTopLevelThisParserPlugin.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+"use strict";
+
+const ConstDependency = require("./ConstDependency");
+
+class HarmonyTopLevelThisParserPlugin {
+	apply(parser) {
+		parser.hooks.expression.for("this").tap("HarmonyTopLevelThisParserPlugin", node => {
+			const module = parser.state.module;
+			const isHarmony = !!(module.buildMeta && module.buildMeta.exportsType);
+			if(isHarmony && parser.scope.topLevelScope) {
```
Comment (Kovensky) [inline lib/dependencies/HarmonyTopLevelThisParserPlugin.js]: Checking `parser.scope.topLevelScope` first might save some time

`if (!parser.scope.topLevelScope) { return; }`

File: lib/Parser.js [hunk 6]
```diff
@@ -624,10 +624,16 @@ class Parser extends Tapable {
 	}
 
 	walkMethodDefinition(methodDefinition) {
+		// TODO: Use `inScope`
```
Comment (sokra) [inline lib/Parser.js]: Use `inScope`

File: lib/dependencies/HarmonyTopLevelThisParserPlugin.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+				const dep = new ConstDependency("undefined", node.range);
```
Comment (sokra) [inline lib/dependencies/HarmonyTopLevelThisParserPlugin.js]: I guess there is a 3rd argument to ConstDependency

File: lib/Parser.js (hunk 6 above)
Comment (ooflorent) [inline lib/Parser.js]: Actually it's not really required. When the parser visits `methodDefinition.value` it calls `inScope` which will change `topLevelScope` because the value is a `FunctionDeclaration`. The tricky case is when class properties will land in the spec because of arrow functions.

**tl;dr**: Currently it's not required to change `topLevelScope` here but it may be mandatory to properly handle the future syntax. Maybe this should be moved to the class definition visitor.

What's your thoughts on this?

File: lib/dependencies/HarmonyTopLevelThisParserPlugin.js (hunk 7 above)
Comment (ooflorent) [reply lib/dependencies/HarmonyTopLevelThisParserPlugin.js]: I don't need it right?

```js
constructor(expression, range, requireWebpackRequire) {
```

Comment (sokra) [reply lib/dependencies/HarmonyTopLevelThisParserPlugin.js]: yes you can pass `false`. The default is `true`

File: lib/Parser.js (hunk 6 above)
Comment (sokra) [reply lib/Parser.js]: Ah I see, then remove all changes in walkMethodDefinition.

ClassProperty will probably a separate AST node type next to MethodDefinition.

Maybe like this: https://astexplorer.net/#/gist/732b089f326bb5ad9260decf2656b7c0/eb3cb03930770076e08d514887930b7960309d24

File: lib/dependencies/HarmonyTopLevelThisParserPlugin.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+class HarmonyTopLevelThisParserPlugin {
+	apply(parser) {
+		parser.hooks.expression.for("this").tap("HarmonyTopLevelThisParserPlugin", node => {
+			if(!parser.scope.topLevelScope)
+				return;
+			const module = parser.state.module;
+			const isHarmony = !!(module.buildMeta && module.buildMeta.exportsType);
+			if(isHarmony) {
+				const dep = new ConstDependency("(void 0)", node.range, false);
```
Comment (sokra) [inline lib/dependencies/HarmonyTopLevelThisParserPlugin.js]: Replace `(void 0)` with `undefined`.

Inserting brackets is not safe because of ASI.

``` js
something
this
```

After replacing and ASI this will call the `something` variable. `something(void 0)`

Comment (ooflorent) [reply lib/dependencies/HarmonyTopLevelThisParserPlugin.js]: Fixed.

**Comment by sokra** [comment]
Thanks
